capsules/src/timer.rs

```diff
@@ -91,54 +112,89 @@ impl<'a, A: Alarm> Driver for TimerDriver<'a, A> {
                 0 /* check if present */ => (ReturnCode::SUCCESS, false),
                 4 /* capture time */ => {
                     let curr_time: u32 = self.alarm.now();
-                    (ReturnCode::SuccessWithValue { value: curr_time as usize }, true)
+                    (ReturnCode::SuccessWithValue { value: curr_time as usize },
+                     false)
                 },
                 3 /* Stop */ => {
-                    if td.interval > 0 {
-                        td.interval = 0;
-                        td.t0 = 0;
-                        let num_armed = self.num_armed.get();
-                        self.num_armed.set(num_armed - 1);
-                        if num_armed == 1 {
-                            self.alarm.disable();
-                            (ReturnCode::SUCCESS, false)
-                        } else {
-                            (ReturnCode::SUCCESS, true)
+                    match td.expiration {
+                        Expiration::Disabled => {
+                            // Request to stop when already stopped
+                            (ReturnCode::EINVAL, false)
+                        },
+                        _ => {
+                            td.expiration = Expiration::Disabled;
+                            td.t0 = 0;
+                            let num_armed = self.num_armed.get();
+                            self.num_armed.set(num_armed - 1);
+                            if num_armed == 1 {
+                                self.alarm.disable();
```

Comment:
I did a little double-take here looking over this. It was really confusing see `num_armed` as non-zero leading to disabling the timer. Could this be re-written as something like:

```rust
let new_num_armed = self.num_armed.get() - 1;
self.num_armed.set(new_num_armed);
if new_num_armed == 0 {
```

This avoids needing to remember to add a `-1` everywhere that `num_armed` is referenced